external/loaders/tests/test__config.py

```diff
@@ -0,0 +1,176 @@
+import tempfile
+import os
+import yaml
+import dataclasses
+
+
+@contextlib.contextmanager
+def mapper_context():
+    original_functions = {**loaders._config.MAPPER_FUNCTIONS}
+    loaders._config.MAPPER_FUNCTIONS.clear()
+    yield
+    loaders._config.MAPPER_FUNCTIONS.clear()
+    loaders._config.MAPPER_FUNCTIONS.update(original_functions)
+
+
+@contextlib.contextmanager
+def batches_context():
+    original_functions = {**loaders._config.BATCHES_FUNCTIONS}
+    loaders._config.BATCHES_FUNCTIONS.clear()
+    yield
+    loaders._config.BATCHES_FUNCTIONS.clear()
+    loaders._config.BATCHES_FUNCTIONS.update(original_functions)
+
+
+def test_load_mapper():
+    with mapper_context():
+        mock_mapper = {"key": xr.Dataset()}
+        mock_mapper_function = unittest.mock.MagicMock(return_value=mock_mapper)
+        mock_mapper_function.__name__ = "mock_mapper_function"
+        loaders._config.register_mapper_function(mock_mapper_function)
+        assert "mock_mapper_function" in loaders._config.MAPPER_FUNCTIONS
+        data_path = "test/data/path"
+        mapper_kwargs = {"arg1": "value1", "arg2": 2}
+        config = loaders._config.MapperConfig(
+            data_path=data_path,
+            mapper_function="mock_mapper_function",
+            mapper_kwargs=mapper_kwargs,
+        )
+        result = config.load_mapper()
+        assert result is mock_mapper
+        mock_mapper_function.assert_called_once_with(data_path, **mapper_kwargs)
+
+
+def test_expected_mapper_functions_exist():
+    # this test exists to ensure we don't accidentally remove functions we
+    # currently use in configs, if you are deleting an option we no longer use
+    # you can delete it here
+    for expected in (
+        "open_nudge_to_obs",
+        "open_nudge_to_fine",
+        "open_fine_res_apparent_sources",
+        "open_high_res_diags",
+        "open_fine_resolution_nudging_hybrid",
+        "open_nudge_to_fine_multiple_datasets",
+    ):
+        assert expected in loaders._config.MAPPER_FUNCTIONS
+
+
+def test_expected_batches_functions_exist():
+    # this test exists to ensure we don't accidentally remove functions we
+    # currently use in configs, if you are deleting an option we no longer use
+    # you can delete it here
+    for expected in (
+        "batches_from_geodata",
+        "batches_from_serialized",
+        "diagnostic_batches_from_geodata",
+    ):
+        assert expected in loaders._config.BATCHES_FUNCTIONS
+
+
+def test_load_batches():
+    with batches_context():
+        mock_batches = [xr.Dataset()]
+        mock_batches_function = unittest.mock.MagicMock(return_value=mock_batches)
+        mock_batches_function.__name__ = "mock_batches_function"
+        loaders._config.register_batches_function(mock_batches_function)
+        assert "mock_batches_function" in loaders._config.BATCHES_FUNCTIONS
+        data_path = "test/data/path"
+        variables = ["var1"]
+        batches_kwargs = {"arg1": "value1", "arg2": 2}
```

Comment:
I reduced duplicated code here quite a bit by pulling the mocking step into the helper.

I was at least able to do this for the two "batches from mapper" test functions, since there are two functionalities (load_batches and load_mapper) that require exactly the same init. For the other two tests, the inits need slight modifications so I left them in place, rather than make the helper too complicated.